src/TRestGeant4ParticleSource.cxx

```diff
@@ -42,51 +47,316 @@ TRestGeant4ParticleSource::~TRestGeant4ParticleSource() {
-    if (GetAngularDistType() == "TH1D") {
-        metadata << "Angular distribution filename : " << GetAngularFilename() << endl;
-        metadata << "Angular histogram name  : " << GetAngularName() << endl;
+    metadata << "Particle Source Name: " << GetParticleName() << endl;
+    if (fParticlesTemplate.size() > 0 || fGenFilename != "") {
+        metadata << "Generator file: " << GetGenFilename() << endl;
+        metadata << "Stored templates: " << fParticlesTemplate.size() << endl;
+        metadata << "Particles: ";
+        for (auto p : fParticles) metadata << p.GetParticleName() << ", ";
+        metadata << endl;
+    } else {
+        metadata << "Charge : " << GetParticleCharge() << endl;
+        metadata << "Angular distribution type : " << GetAngularDistType() << endl;
+        if (GetAngularDistType() == "TH1D") {
+            metadata << "Angular distribution filename : " << GetAngularFilename() << endl;
+            metadata << "Angular histogram name  : " << GetAngularName() << endl;
+        }
+        metadata << "Direction : (" << GetDirection().X() << "," << GetDirection().Y() << ","
+                 << GetDirection().Z() << ")" << endl;
+        metadata << "Energy distribution : " << GetEnergyDistType() << endl;
+        if (GetEnergyDistType() == "TH1D") {
+            metadata << "Energy distribution filename : " << GetSpectrumFilename() << endl;
+            metadata << "Energy histogram name  : " << GetSpectrumName() << endl;
+        }
+        if (GetEnergyRange().X() == GetEnergyRange().Y())
+            metadata << "Energy : " << GetEnergy() << " keV" << endl;
+        else
+            metadata << "Energy range : (" << GetEnergyRange().X() << "," << GetEnergyRange().Y() << ") keV"
+                     << endl;
     }
-    metadata << "Direction : (" << GetDirection().X() << "," << GetDirection().Y() << ","
-             << GetDirection().Z() << ")" << endl;
-    metadata << "Energy distribution : " << GetEnergyDistType() << endl;
-    if (GetEnergyDistType() == "TH1D") {
-        metadata << "Energy distribution filename : " << GetSpectrumFilename() << endl;
-        metadata << "Energy histogram name  : " << GetSpectrumName() << endl;
+}
+
+TRestGeant4ParticleSource* TRestGeant4ParticleSource::instantiate(std::string model) {
+    if (model == "" || model == "geant4" || model.find(".dat") != -1) {
+        // use default generator
+        return new TRestGeant4ParticleSource();
+    } else {
+        // use specific generator
+        // naming convension: TRestGeant4ParticleSourceXXX
+        // currently supported generator: decay0
+        // in future we may add wrapper of generators: cry(for muon), pythia(for HEP), etc.
+        model[0] = *REST_StringHelper::ToUpper(std::string(&model[0], 1)).c_str();
+        TClass* c = TClass::GetClass(("TRestGeant4ParticleSource" + model).c_str());
+        if (c != NULL)  // this means we have the package installed
```

Comment:
`if(c)`

Reply:
Using `if (c != NULL)` or `if (c)` is just a matter of coding perference. Someone likes the former since it is more expilict and others thinks the later is more brief. Does REST have such style convention?

Reply:
I believe we do not state such convention, except what it is inside Google c++ style guide

https://google.github.io/styleguide/cppguide.html

We give the reference at our CONTRIBUTING guide

https://github.com/rest-for-physics/framework/blob/master/CONTRIBUTING.md

Of course, I never read that carefully, and I believe the only way we manage to follow that style is by integrating tools such as `clang-tidy`. Anyhow, thats a good point, and I will try to follow @juanangp suggestion

Reply:
I think we should avoid the use of `NULL`, we should move to `nullptr` as the code is C++ oriented

The use of `if (c != nullptr)` or `if (c)` it is user preference, although I think we should have an style convention in the repository.

Reply:
Ok, done here. Indeed, we have a codig style, but as I mentioned it is hard everyone follows, adapts to it. We need a tool such as clang-tidy helping us on that task.